Userland/Libraries/LibWeb/Painting/PaintableBox.cpp

```diff
@@ -700,6 +703,21 @@ Optional<HitTestResult> PaintableWithLines::hit_test(Gfx::FloatPoint const& posi
                 return HitTestResult { *fragment.layout_node().paintable(), fragment.text_index_at(position.x()) };
             }
 
+            // Don't test input text elements for text cursor hit
+            // FIXME: There should be a better way to know if a fragment comes from an input element, perhaps stored on the fragment
+            if (fragment.layout_node().dom_node()->first_ancestor_of_type<DOM::ShadowRoot>() != nullptr) {
+                continue;
+            }
+            // Don't test text within button elements for text cursor hit
+            if (fragment.layout_node().dom_node()->first_ancestor_of_type<HTML::HTMLButtonElement>() != nullptr) {
+                continue;
+            }
+            // FIXME: Text from a label element should be tested for text cursor hit, but currently that breaks the algorithm below.
+            //  From a dump of the layout tree, it looks like labels create their own separate line boxes.
+            if (fragment.layout_node().dom_node()->first_ancestor_of_type<HTML::HTMLLabelElement>() != nullptr) {
+                continue;
+            }
```

Comment:
This feels a bit *too* hackish. By the time we've transformed from DOM tree -> layout tree -> paint tree, it's sad to suddenly have various based on the original DOM node.

Can we instead make these behaviors governed by flags/state in the paint tree somehow?

Reply:
I totally agree that it's too hackish. From my limited understanding of the architecture, putting a flag on the fragment in the layout tree seems like a natural better approach. If we try to make selection work only on the paint tree, won't we have lost too much information about how fragments are layed out? I suppose we could calculate everything from the paint boxes alone; is that better architecturally?

I will remove this commit from the pull request, and work on a better way of handling this. Thanks for your feedback :)

Reply:
Conceptually, text selection should be based on visible content, and the paint tree is basically the representation of that. Note that the layout tree still hangs around, and every paintable can always reach into its layout node for information.

The important thing is that selection can't be based on the DOM alone, since that doesn't include generated content like pseudo elements.